src/mgds/pipelineModules/DropTags.py

```diff
@@ -0,0 +1,131 @@
+            delimiter_in_name: str,
+            regex_enabled_in_name: str,     #whether to process special tags with regex
+            keep_tags_count_in_name: str,
+            text_out_name: str,
+    ):
+        super(DropTags, self).__init__()
+        self.text_in_name = text_in_name
+        self.enabled_in_name = enabled_in_name
+        self.probability_in_name = probability_in_name
+        self.dropout_mode_in_name = dropout_mode_in_name
+        self.special_tags_in_name = special_tags_in_name
+        self.special_tag_mode_in_name = special_tag_mode_in_name
+        self.delimiter_in_name = delimiter_in_name
+        self.regex_enabled_in_name = regex_enabled_in_name
+        self.keep_tags_count_in_name = keep_tags_count_in_name
+        self.text_out_name = text_out_name
+
+    def length(self) -> int:
+        return self._get_previous_length(self.text_in_name)
+
+    def get_inputs(self) -> list[str]:
+        return [self.text_in_name, self.enabled_in_name, self.probability_in_name, self.dropout_mode_in_name, self.special_tags_in_name, self.special_tag_mode_in_name, self.delimiter_in_name, self.keep_tags_count_in_name]
+
+    def get_outputs(self) -> list[str]:
+        return [self.text_out_name]
+
+    def get_item(self, variation: int, index: int, requested_name: str = None) -> dict:
+        text = self._get_previous_item(variation, self.text_in_name, index)
+        delimiter = self._get_previous_item(variation, self.delimiter_in_name, index)
+        keep_tags_count = self._get_previous_item(variation, self.keep_tags_count_in_name, index)
+        enabled = self._get_previous_item(variation, self.enabled_in_name, index)
+        probability = self._get_previous_item(variation, self.probability_in_name, index)
+        dropout_mode = self._get_previous_item(variation, self.dropout_mode_in_name, index)
+        special_tag_mode = self._get_previous_item(variation, self.special_tag_mode_in_name, index)
+        special_tags = self._get_previous_item(variation, self.special_tags_in_name, index)
+        regex_enabled = self._get_previous_item(variation, self.regex_enabled_in_name, index)
+        rand = self._get_rand(variation, index)
+
+        if enabled and (probability > 0):
+            #convert inputs to lists and set up final output list (pruned) and white/blacklist (special)
+            tags = [tag.strip() for tag in text.split(delimiter)]
+            keep_tags = tags[:keep_tags_count]
+            dropout_tags = tags[keep_tags_count:]
+            pruned_tags = []
+            special_tags_prelist = []
+            special_tags_list = []
+
+            #convert special_tags to list depending on whether it's a newline-separated csv/txt file or a delimiter-separated string
+            if (special_tags.endswith(".txt") and os.path.isfile(special_tags)):
+                with open(special_tags) as special_tags_file:
```

Comment:
Might be a good idea to add dome kind of in-memory cache to this. Reading and parsing a txt/csv file on every training step will add a bit too much overhead.

Reply:
Thanks for the suggestion, I'll have to test this with some larger datasets and see how much of a slowdown it causes. I mostly included that part to make it easier to reuse large and complex white/blacklists between multiple concepts, but if you want to exclude that for now and just use the text field only, I can look into it more and submit it in the future.

Reply:
Did some research and some time comparisons between loading special tags from file vs giving it a string directly, looping the main block of the script through a list of ~1200 captions. Loading special tags from a file took around 0.12 s to complete, compared to 0.02 s for string. Changing the file loading section into a function with /@functools.lru_cache closed the difference between them, however I'm not sure if the cache will persist each time the real function is called and I don't know how to do that without writing to disk. I can add it to the PR anyway since it didn't seem to hurt anything at least.

Enabling regex also made it slower but there may not be a way to improve that, caching the regex patterns didn't help. In any case it seems like a difference of tenths of a millisecond for a single iteration so I don't think it's going to affect training speed. I did also try running some short trainings comparing string/txt file, as well as regex on/off, and the completion time was not noticeably different. My system is not the strongest so it may be slightly more noticeable with faster GPUs.

Reply:
I'm not even sure if a LRU cache is needed. I would have just added a dict that to map a filename to the parsed content. unless someone adds millions of different files, that create a memory issue, and it also ensures that every file only creates a single cache miss

Reply:
Alright, maybe I misunderstood what you were asking for. There would only ever be one "special tag" file per concept right now, so memory issues would be unlikely. I really don't have much experience with Python beyond simple single-purpose scripts so I don't know what best practices would be for caching a handful of text files between multiple calls to the same script.

I think the main thing that makes this "slow" is the fact that it's only doing one caption->one variation each time it runs, and loading/interpreting the special list from scratch each time it loads a new caption or generates a new variation. That's the way the "shuffle" module works and it's what I used as a reference for the structure of both of these. It wouldn't be difficult to change it to return a list of n variations of a single caption, but it seems like that would require significant changes to the way all modules are handled in the rest of the project.